beacon_node/eth2_libp2p/src/discovery/mod.rs

```diff
@@ -519,63 +567,119 @@ impl<TSpec: EthSpec> Discovery<TSpec> {
         let eth2_fork_predicate = move |enr: &Enr| enr.eth2() == Ok(enr_fork_id.clone());
 
         // General predicate
-        let predicate: Box<dyn Fn(&Enr) -> bool + Send> = match &query {
-            QueryType::FindPeers => Box::new(eth2_fork_predicate),
-            QueryType::Subnet { subnet_id, .. } => {
-                // build the subnet predicate as a combination of the eth2_fork_predicate and the
-                // subnet predicate
-                let subnet_predicate = subnet_predicate::<TSpec>(*subnet_id, &self.log);
-                Box::new(move |enr: &Enr| eth2_fork_predicate(enr) && subnet_predicate(enr))
-            }
-        };
+        let predicate: Box<dyn Fn(&Enr) -> bool + Send> =
+            Box::new(move |enr: &Enr| eth2_fork_predicate(enr) && additional_predicate(enr));
 
         // Build the future
         let query_future = self
             .discv5
             .find_node_predicate(random_node, predicate, target_peers)
-            .map(|v| QueryResult(query, v));
+            .map(|v| QueryResult(grouped_query, v));
 
         // Add the future to active queries, to be executed.
         self.active_queries.push(Box::pin(query_future));
     }
 
     /// Drives the queries returning any results from completed queries.
-    fn poll_queries(&mut self, cx: &mut Context) -> Option<(Option<Instant>, Vec<Enr>)> {
+    fn poll_queries(&mut self, cx: &mut Context) -> Option<HashMap<PeerId, Option<Instant>>> {
         while let Poll::Ready(Some(query_future)) = self.active_queries.poll_next_unpin(cx) {
             match query_future.0 {
-                QueryType::FindPeers => {
+                GroupedQueryType::FindPeers => {
                     self.find_peer_active = false;
                     match query_future.1 {
                         Ok(r) if r.is_empty() => {
                             debug!(self.log, "Discovery query yielded no results.");
                         }
                         Ok(r) => {
                             debug!(self.log, "Discovery query completed"; "peers_found" => r.len());
-                            return Some((None, r));
+                            let mut results: HashMap<PeerId, Option<Instant>> = HashMap::new();
```

Comment:
we should add to `cached_enrs` here as well

Reply:
done 👍